CSHARP-2672: Fix Snappy compression

vincentkam opened the pull request:
https://jira.mongodb.org/browse/CSHARP-2672
Rebased to include https://github.com/rstam/mongo-csharp-driver/pull/128

vincentkam commented on evergreen/evergreen.yml:
```diff
@@ -218,7 +218,7 @@ functions:
           export FLE_AWS_SECRET_ACCESS_KEY=${FLE_AWS_SECRET_ACCESS_KEY}
           ${PREPARE_SHELL}
-          AUTH=${AUTH} SSL=${SSL} MONGODB_URI="${MONGODB_URI}" TOPOLOGY=${TOPOLOGY} OS=${OS} evergreen/run-tests.sh
+          AUTH=${AUTH} SSL=${SSL} MONGODB_URI="${MONGODB_URI}" TOPOLOGY=${TOPOLOGY} OS=${OS} COMPRESSOR=${COMPRESSOR} evergreen/run-tests.sh
```
This line is needed to pass the compressor variable to `run-tests.sh`

vincentkam commented on evergreen/evergreen.yml:
```diff
@@ -483,12 +483,12 @@ buildvariants:
-#  tags: ["tests-variant"]
-#  tasks:
-#     - name: "test"
+- matrix_name: "tests-snappy-compression"
```
Re-enabling snappy compression tests

vincentkam commented on evergreen/run-tests.sh:
```diff
@@ -14,6 +14,7 @@ AUTH=${AUTH:-noauth}
 SSL=${SSL:-nossl}
 MONGODB_URI=${MONGODB_URI:-}
 TOPOLOGY=${TOPOLOGY:-server}
+COMPRESSOR=${COMPRESSOR:-none}
```
Set the compressor to "none" if the env var does not exist

vincentkam commented on evergreen/run-tests.sh:
```diff
@@ -30,6 +31,15 @@ provision_ssl () {
 
+provision_compressor () {
+    uri_environment_variable_name=$1
+    if [[ "${!uri_environment_variable_name}" =~ "/?" ]]; then
```
I chose a different approach here than `provision_ssl`. `provision_ssl` searches for a specific combination of options that results in a URI that doesn't have additional options e.g. a URI of the form `mongodb://user:pass@localhost/`. Determing this combination of options is nontrivial, so instead, I've chosen to use a regex to search for `/?` in the URI. The downside of this the script will only run on bash instead of a generic POSIX compliant shell (e.g. whatever may be at `/bin/sh`), but this shouldn't be a problem since we were using `#/bin/bash` before anyways.

vincentkam replied:
This function adds the compressor to the URI, so this'll enable `zlib` testing too

DmitryLukyanov commented on src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs:
```diff
@@ -27,28 +26,104 @@ internal class SnappyCompressor : ICompressor
+            stream.ReadBytes(bytesRead, offset: 0, count: bytesRead.Length, CancellationToken.None);
+            return bytesRead[0] | (bytesRead[1] << 8) | (bytesRead[2] << 16) | (bytesRead[3] << 24);
+        }
+
```
not needed empty line?

vincentkam replied:
Fixed!

DmitryLukyanov commented on src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs:
```diff
@@ -27,28 +26,104 @@ internal class SnappyCompressor : ICompressor
+            var compressedBytes = new byte[maxSize];
+            var compressedSize = SnappyCodec.Compress(uncompressedMessage, 0, uncompressedSize, compressedBytes, 0);
+            output.Write(compressedBytes, 0, compressedSize);
+
```
not necessary empty line?

vincentkam replied:
Fixed!

vincentkam commented:
Ready for review!

vincentkam commented on evergreen/run-tests.sh:
```diff
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/env bash
```
This makes is so that the path is searched for `bash`. This is especially useful on when the user may be using newer version of bash that's not stored at `/bin/bash` (e.g. homebrew bash on MacOS)

vincentkam commented:
Sorry, had to fix one more thing: now ready for review!

rstam commented on src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs:
```diff
@@ -27,28 +27,89 @@ internal class SnappyCompressor : ICompressor
+            var maxSize = SnappyCodec.GetMaxCompressedLength(uncompressedSize);
+            var compressedBytes = new byte[maxSize];
+            var compressedSize = SnappyCodec.Compress(uncompressedMessage, 0, uncompressedSize, compressedBytes, 0);
+            output.Write(compressedBytes, 0, compressedSize);
```
I don't like the idea of *reading* from `output`.

Will this work instead:

        public void Compress(Stream input, Stream output)
        {
    #if NET452 || NETSTANDARD2_0
            var uncompressedSize = (int)(input.Length - input.Position);
            var uncompressedBytes = new byte[uncompressedSize]; // does not include uncompressed message headers
            input.ReadBytes(uncompressedBytes, offset: 0, count: uncompressedSize, CancellationToken.None);
            var maxCompressedSize = SnappyCodec.GetMaxCompressedLength(uncompressedSize);
            var compressedBytes = new byte[maxCompressedSize];
            var compressedSize = SnappyCodec.Compress(uncompressedBytes, 0, uncompressedSize, compressedBytes, 0);
            output.Write(compressedBytes, 0, compressedSize);
    #else
            throw new NotSupportedException();
    #endif
        }

Also... let's try and reserve word "message" for when talking about the entire message, including headers.

vincentkam replied:
Re: "message": ack. 
Your suggestion is definitely cleaner and works. Fixed!

rstam commented on src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs:
```diff
@@ -27,28 +27,89 @@ internal class SnappyCompressor : ICompressor
+            var compressedMessage = new byte[messageLength];
+            input.ReadBytes(compressedMessage, offset: 0, count: messageLength, CancellationToken.None);
+            var decompressedMessage = SnappyCodec.Uncompress(compressedMessage);
+            output.Write(decompressedMessage, offset: 0, count: decompressedMessage.Length);
```
I don't like the idea of *reading* from `output`.

Will this work instead:

            public void Decompress(Stream input, Stream output)
            {
    #if NET452 || NETSTANDARD2_0
                var compessedSize = (int)(input.Length - input.Position);
                var compressedBytes = new byte[compessedSize];
                input.ReadBytes(compressedBytes, offset: 0, count: compessedSize, CancellationToken.None);
                var decompressedBytes = SnappyCodec.Uncompress(compressedBytes);
                output.Write(decompressedBytes, offset: 0, count: decompressedBytes.Length);
    #else
                throw new NotSupportedException();
    #endif
            }

vincentkam replied:
Your suggestion is definitely cleaner and works. Fixed!

rstam commented on src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs:
```diff
@@ -27,28 +27,89 @@ internal class SnappyCompressor : ICompressor
+        private const int OpCodeSize = sizeof(int);
+        private const int OriginalOpCodeSize = OpCodeSize;
+        private const int UncompressedSizeSize = sizeof(int);
+        private const int CompressorTypeSize = sizeof(byte); // compressorId in spec
```
These constants might no longer be needed depending on what changes you make below.

vincentkam replied:
Ack, fixed

rstam commented on src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs:
```diff
@@ -27,28 +27,89 @@ internal class SnappyCompressor : ICompressor
+        /// precondition.
+        /// The precondition for <paramref name="output"/> mirrors the precondition for <see cref="Decompress"/>'s input
+        /// precondition.
+        /// </remarks>
```
These comments might no longer be correct depending on what changes you make below.

vincentkam replied:
Hmm.. I think the preconditions technically still hold true, despite us changing the implementation, and leaving them there makes it easier to reason about the state of the stream (as the preconditions precisely define what the state the stream is in). I'd strongly advocate leaving them in as this was the part that was hardest to wrap my head around when I was investigating and imagine the comments will be useful in the future if we return to this section

rstam replied:
They are no longer preconditions. They only became preconditions when you were seeking backwards into the stream to peek at the data that you assumed had been planted there.

I would like to see these comments removed.

rstam replied:
Think of it this way: if we had a file called `SnappyCompressorTests` (not sure why we don't) you would NOT need to write any headers to the stream before calling the `Compress` method.

So this is NOT a precondition.

vincentkam replied:
Applied changes as discussed offline!

rstam commented on src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs:
```diff
@@ -27,28 +27,89 @@ internal class SnappyCompressor : ICompressor
+        /// precondition.
+        /// The precondition for <paramref name="output"/> mirrors the precondition for <see cref="Compress"/>'s input
+        /// precondition.
+        /// </remarks>
```
These comments might no longer be correct depending on what changes you make below.

vincentkam replied:
Applied changes as discussed offline!

rstam commented on src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs:
```diff
@@ -27,28 +27,89 @@ internal class SnappyCompressor : ICompressor
 #endif
         }
+
+        private int ReadInt32(Stream stream)
```
Might no longer be needed.

vincentkam replied:
Fixed!

DmitryLukyanov commented on tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs:
```diff
@@ -3611,6 +3611,9 @@ private MongoCollection<BsonDocument> GetCollection(MongoServer server)
 
         private MongoServerInstance GetPrimary(MongoServer server)
         {
+            SpinWait.SpinUntil(
```
why is this change required?

vincentkam replied:
I think it's because `TestFindAndRemoveWithMaxTime` spins up a new `MongoClient` by passing in `retryWrites: false`, resulting in a race condition where we try to get the primary before the topology has finished initializing. 

See: https://jira.mongodb.org/browse/CSHARP-2755 and its PR for a bit more detail
(Robert's also looking more deeply into it).

vincentkam replied:
This change has been removed and replaced with Robert's "Equals" fix

DmitryLukyanov commented on src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs:
```diff
@@ -27,25 +27,57 @@ internal class SnappyCompressor : ICompressor
-            {
-                snappyStream.CopyTo(output);
-            }
+#if NET452 || NETSTANDARD2_
```
Is it a typo: `2_`?

vincentkam replied:
Thanks Dima! Fixed!

rstam commented on Release Notes/Release Notes v2.9.2.md:
```diff
@@ -1,7 +1,8 @@
+# .NET Driver Version 2.9.2 Release Notes
 
-The main new features in 2.9.2 are:
+This is a patch release that fixes several bugs reported since 2.9.0 was released.
```
Since 2.9.1 (not 2.9.0)?

vincentkam replied:
Fixed!

rstam commented on src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs:
```diff
@@ -27,25 +27,57 @@ internal class SnappyCompressor : ICompressor
-            {
-                snappyStream.CopyTo(output);
-            }
+            var compessedSize = (int) (input.Length - input.Position);
```
Type, should be `compressedSize` (missing an `r`)

vincentkam replied:
Fixed!

vincentkam commented:
Ready for re-review! Primarily, I've changed the evergreen.yml file to drop 3.4 from testing, and updated the release notes with some snappy information. I know that typically we've done the release notes updating as a separate PR, but this seemed like an easy way to "amortize" some of the effort

vincentkam commented:
Ready for re^2-review!

vincentkam commented:
Rebased, squashed, date updated and pushed!
